### ISSUE TITLE
Upgrade Client App to React 18

### ISSUE THREAD

**Issue by firasrg** [spring-petclinic/spring-petclinic-reactjs#25]
I would like to propose upgrading [the client app](https://github.com/spring-petclinic/spring-petclinic-reactjs/tree/master/client) of this project to [React v18](https://react.dev/blog/2022/03/29/react-v18). 

## Problem
The current version of React used in the project may lack some of the latest features and optimizations introduced in the modern versions. Additionally, the application may face compatibility issues with the latest React ecosystem, libraries, and tooling. 

Upgrading to modern versions will address these problems and provide developers with access to the most recent advancements in React .

## Solution
1. Upgrade to Modern React Version (includes verification of dependencies compatibility).
2. Additional improvements/features:
   - Evaluate areas that would need new features or improvements.
   - Discuss and plan the addition of features that align with the goals of the Spring PetClinic project.

### Potential new features
1. Enhanced data visualization
2. Improved user experience
3. Any functionality that can add value to the application.

## Benefits
- Improved development experience and productivity.
- Ensuring compatibility with the evolving React ecosystem.
- Potential to enhance user experience and provide additional value to the Spring PetClinic application.

## References
- Link to React documentation: [React](https://react.dev/)

I believe that upgrading and adding new features will enhance the development experience. This will enable developers to work with the latest technologies while providing additional value to the users of the Spring PetClinic application.

Please let me know if you have any questions or if there's any further information I can provide to support this enhancement proposal.

**Comment by michaelisvy** [comment]
I agree that it needs an upgrade, and at this stage I haven't managed to start the application successfully (it might be due to my version of Node.js being too recent).

**Comment by arey** [comment]
Like @michaelisvy , an update of the React front and Spring Boot backend would be a good thing.  Since you seem motivated and are motivated @firasrg , you have free rein and could work on it. I think @nilshartmann  will agree too.

**Comment by nilshartmann** [comment]
Hi all!

I think that's a very good idea to upgrade this. Unfortunatly I hadn't found the time yet, so I'm more than happy if anyone else would support.

Tbh I think "upgrade" in this case is more or less a re-write 😬

Does it make sense, before starting with the upgrade to could discuss a tech stack that we want to use?

For example: 
- should we go with create-react-app or vite? 
- use TanStack Query for data fetching?
- CSS: "plain", modules, css-in-js, Tailwind, ... ?
- should we implement frontend tests? should they run in GitHub actions? 🙀

**Comment by firasrg** [comment]
Hi all

@michaelisvy I got the client app working without backend and data, just UI.

@nilshartmann yeah it needs to be rewritten, I have a good experience with React I can provide a good initiative on that. As for techs, I imagine we would need CRA (create react app) for sure, React Query for datafetching yes! Tailwind CSS or something similar for styling and layouts. 

As for tests, I think : [Jest ](https://jestjs.io/fr/)+ [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/)

I have a good background with these tools **and some others**.

**Comment by nilshartmann** [comment]
@firasrg not sure, if `create-react-app` is still a good idea. That project seems (currently) more or less abandoned while more and more people tend to use [Vite](https://vitejs.dev/guide/#scaffolding-your-first-vite-project). That would probably mean [vitest](https://vitest.dev/) instead of jest, but that should also work with RTL.

**Comment by firasrg** [comment]
TBH I don't know Vite, but based on [a first comparison](https://blog.logrocket.com/vite-3-vs-create-react-app-comparison-migration-guide), it seems a good tool that brings better performance over CRA. But I think for beginners and intermediate levels, CRA would be more suitable as it's mature, popular and easier to use.  As PetClinic project is intended to be used by new people or immigrants, I think it would be better to start with CRA.

It's up to you

**Comment by firasrg** [comment]
> ~~I think it would be better to start with CRA~~.

I just got an information from trusted React community members that [CRA](https://github.com/facebook/create-react-app) is no more maintained. I agree with what you said

**Comment by michaelisvy** [comment]
hello, in case the work can be split, I'll be happy to help out (either for React part or for the Backend part)

**Comment by nilshartmann** [comment]
> hello, in case the work can be split, I'll be happy to help out (either for React part or for the Backend part)

Hi @michaelisvy,

thank you very much!!!

I think it makes sense to consolidate the backend and frontend issue, so please find my ideas here: https://github.com/spring-petclinic/spring-petclinic-reactjs/issues/26#issuecomment-1586740581

**Comment by firasrg** [comment]
Alright@nilshartmann , in this case I close this issue.
